examples/post_training_quantization/openvino/tiny_gpt2/main.py

```diff
@@ -0,0 +1,64 @@
+from optimum.intel.openvino import OVModelForCausalLM
+from transformers import AutoTokenizer
+
+import nncf
+
+GENERATION_LENGTH = 20
+
+
+def fix_ov_model_names_duplicates(ov_model):
+    names = set()
+    for op in ov_model.get_ops():
+        friendly_name = op.get_friendly_name()
+        while True:
+            if friendly_name not in names:
+                break
+            friendly_name += "_"
+        names.add(friendly_name)
+        op.set_friendly_name(friendly_name)
+
+
+model_id = "hf-internal-testing/tiny-random-gpt2"
+# model_id = "hf-internal-testing/tiny-random-GPTNeoModel"
+# model_id = "hf-internal-testing/tiny-random-GPTNeoXForCausalLM"
+
+tokenizer = AutoTokenizer.from_pretrained(model_id)
+tokens = tokenizer("This is a sample input", return_tensors="pt")
+
+model_with_pkv = OVModelForCausalLM.from_pretrained(model_id, export=True, use_cache=True)
+
+
+def set_ov_model_in_hf_model(hf_model, ov_model):
+    hf_model.model = ov_model
+    hf_model.request = ov_model.create_infer_request()
+
+
+def get_custom_forward(ov_model, callback_fn):
+    hf_model = model_with_pkv
+    set_ov_model_in_hf_model(hf_model, ov_model)
+
+    def _callback_fn(info):
+        outputs = {k: v for k, v in zip(info["infer_request"].model_outputs, info["infer_request"].outputs)}
+        callback_fn(outputs)
+
+    hf_model.request.set_callback(_callback_fn, {"infer_request": hf_model.request})
+
+    def custom_forward(dataitem):
+        hf_model.generate(**dataitem, min_length=GENERATION_LENGTH, max_length=GENERATION_LENGTH, num_beams=1)
+
+    return custom_forward
+
+
+def transform_fn(data_item):
+    return data_item
+
+
+dataset = nncf.CustomInferenceDataset([tokens] * 10, transform_fn, get_custom_forward)
```

Comment:
I don't think we should make get_custom_forward a part of Dataset API. I propose:
- rename it to `get_forward_fn(model: ov.Model, output_processing_callback: Callable) -> Callable` 
- make it an optional argument of nncf.quantize() API

Reply:
I absolutely agree that it should not be part of Dataset API.

Comments from my side:
I have some concerns about get_forward_fn:

1. `output_processing_callback` is not needed for Torch and Keras TF models. It can confuse developer because they will call `output_processing_callback` that does not do it anything.
2. `signature of output_processing_callback` is not clear for different frameworks.

Proposal:
1. Introduce `get_forward_fn(model: ov.Model) -> Callable` Torch and Keras TF and `get_forward_fn(model: ov.Model, statistic_aggregator: StatisticsAggregator) -> Callable` for OpenVINO, ONNX and TF.

Pros:

- It is addressed to 1 via explicit introduction different signatures for different frameworks because different frameworks collect statistics with using different approaches.

- It is addressed to 2 because methods of a class can be easily documented + sugar from IDE. It also can provide several interfaces to register model output. `statistic_collector.register_model_output(name, tensor) ``statistic_collector.register_model_outputs(outputs: Dict[str, tensor]) `
Request changes:
```
def get_custom_forward(ov_model, statistic_aggregator):
    hf_model = model_with_pkv
    set_ov_model_in_hf_model(hf_model, ov_model)

    def _callback_fn(info):
        outputs = {k.key.get_any_name(): v.value for k, v in zip(info["infer_request"].model_outputs, info["infer_request"].outputs)}
        statistic_aggregator.register_model_outputs(outputs)
```

2. Introduce a different classes to join framework model and custom forward function for each framework. For example`nncf.OVModelWithCustomForward(model: ov.Model, get_forward_fn: Callable)` for OV

Pros:
- `nncf.quantize` and `nncf.quantize_with_accuracy_control` w/o extending signature
- The class explicitly specified signature of get_forward_fn for framework model.
- Easy reuse in another algorithms
```
ov_model_with_custom_forward = nncf.OVModelWithCustomForward(model_with_pkv.model, get_forward_fn)
quantized_model_with_custom_forward = nncf.quantize(ov_model_with_custom_forward, dataset, subset_size=3)
```

3. IMHO: rename `get_forward_fn` -> `make_forward_fn`

Reply:
To tell the truth, I am still skeptical about the whole approach of collecting recurrent states and how this is applicable to other models. Now, I am looking at the Whisper [notebook](https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/227-whisper-subtitles-generation/227-whisper-subtitles-generation.ipynb) and I would not use this API since it requires much more effort and code rewriting to use the proposed API.